posts/2019/07/uarray-update-api-changes-overhead-and-comparison-to-__array_function__.ipynb

```diff
@@ -7,6 +7,38 @@
     "`uarray` is a generic override framework for objects and methods in Python. Since [my last `uarray` blogpost](https://labs.quansight.org/blog/2019/04/uarray-intro/), there have been plenty of developments, changes to the API and improvements to the overhead of the protocol. Let’s begin with a walk-through of the current feature set and API, and then move on to current developments and how it compares to [`__array_function__`](www.numpy.org/neps/nep-0018-array-function-protocol.html). For further details on the API and latest developments, please see [the API page for `uarray`](https://uarray.readthedocs.io/en/latest/generated/uarray.html). The examples there are doctested, so they will always be current."
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Motivation\n",
+    "### Other array objects\n",
+    "NumPy is a simple, rectangular, dense, and in-memory data store. This is great for some applications but isn't complete on its own. It doesn't encompass every single use-case. The following are examples of array objects available today that have different features and cater to a different kind of audience.\n",
+    "\n",
+    "* Dask is one of the most popular ones. It allows distributed and chunked computation.\n",
+    "* CuPy is another popular one, and allows GPU computation.\n",
+    "* PyData/Sparse is slowly gaining popularity, and is a sparse, in-memory data store.\n",
+    "* XArray includes named dimensions.\n",
+    "* Xnd is another effort to re-write and modernise the NumPy API, and includes support for GPU arrays and ragged arrays.\n",
+    "* Another effort (although with no Python wrapper, only data marshalling) is xtensor.\n",
+    "\n",
+    "Some of these objects can be composed. Namely, Dask both expects and exports the NumPy API, whereas XArray expects the NumPy API. This makes interesting combinations possible, such as distributed sparse or GPU arrays, or even labelled distributed sparse or CPU/GPU arrays.\n",
+    "\n",
+    "Also, there are many other libraries (a popular one being scikit-learn) that need a back-end mechanism in order to be able to support different kinds of array objects. Finally, there is a desire to see SciPy itself gain support for other array objects.\n",
+    "\n",
+    "### `__array_function__` and its limitations\n",
+    "One of my motivations for working on `uarray` were the limitations of the `__array_function__` protocol, defined in [this proposal](https://www.numpy.org/neps/nep-0018-array-function-protocol.html). The limitations are threefold:\n",
+    "\n",
+    "* It can only dispatch on array objects.\n",
+    "* Consequently, it can only dispatch on functions that *accept* array objects.\n",
```

Comment:
There's one other consequence: It can only support _a single backend_ per array type (so, e.g., supporting both PocketFFT and PyFFTW for `numpy.ndarray` input in `scipy.fft` isn't possible).